### PR TITLE
add shortcut for list item ordering

### PR DESCRIPTION
**Pull request by Laserlicht**
for hero and town list:
- add possibility to move with keyboard and mouse
  - hover item with mouse and press PageUp, PageDown, End, Pos1 to order
- make it possible to dismiss hero with Del in list
- fix bug of not updating list when ordering in town screen

### CODE REVIEW

File: client/adventureMap/CList.cpp [hunk 1]
```diff
@@ -301,6 +305,72 @@ void CHeroList::CHeroItem::gesture(bool on, const Point & initialPosition, const
...
+		ony += [=](){ LOCPLINT->cb->dismissHero(hero); };
+		LOCPLINT->showYesNoDialog(CGI->generaltexth->allTexts[22], ony, nullptr);
+		return;
+	}
+
+	if(heroes.size() < 2)
+		return;
+
+	size_t heroPos = vstd::find_pos(heroes, hero);
+	const CGHeroInstance * heroUpper = (heroPos < 1) ? nullptr : heroes.at(heroPos - 1);
```
Comment (nullkiller) [inline client/adventureMap/CList.cpp:345]: Can we use boolean isHeroUpper here instead? Or can heroes.at(index) contain nullptr itself?

File: client/adventureMap/CList.cpp [hunk 2]
```diff
@@ -301,6 +305,72 @@ void CHeroList::CHeroItem::gesture(bool on, const Point & initialPosition, const
...
+		return;
+
+	if(!pos.isInside(GH.getCursorPosition()))
+		return;
+
+	auto & heroes = LOCPLINT->localState->getWanderingHeroes();
+
+	if(key == EShortcut::LIST_HERO_DELETE)
+	{
+		CFunctionList<void()> ony = [=](){ };
```
Comment (nullkiller) [inline client/adventureMap/CList.cpp:335]: What does ony mean?

**Comment by Laserlicht** [comment]
> Perhaps instead of targeting hovered hero/town instead apply shortcuts on selected object? Should simplify code a bit and will make it possible to use these shortcuts without mouse

The problem here is that this shortcuts overlaps with adventure map shortcuts. With the current approach no problem.

File: client/adventureMap/CList.cpp (hunk 1 above)
Comment (Laserlicht) [reply client/adventureMap/CList.cpp:345]: Not sure. Used the same line as for the gesture here. To be safe.

File: client/adventureMap/CList.cpp (hunk 2 above)
Comment (Laserlicht) [reply client/adventureMap/CList.cpp]: I think `on yes`. Also used in another place in VCMI code base like this. Removed. As it doesn't make sense to store lambda in var.

**Comment by IvanSavenko** [comment]
> The problem here is that this shortcuts overlaps with adventure map shortcuts. With the current approach no problem.

If player already moved mouse he might just as well use mouse wheel. 

Perhaps assign to free hotkeys, e.g. Ctrl+PageUp / Ctrl+Up / etc? For me it is weird to have keyboard hotkeys that require mouse positioning.

**Comment by Laserlicht** [comment]
Changed to Ctrl+...

File: client/gui/Shortcut.h [hunk 3]
```diff
@@ -293,5 +293,15 @@ enum class EShortcut
 	SPELLBOOK_TAB_ADVENTURE,
 	SPELLBOOK_TAB_COMBAT,
 
+	LIST_HERO_UP,
+	LIST_HERO_DOWN,
+	LIST_HERO_TOP,
+	LIST_HERO_BOTTOM,
+	LIST_HERO_DELETE,
```
Comment (IvanSavenko) [inline client/gui/Shortcut.h:300]:
```suggestion
	LIST_HERO_DISMISS,
```

**Review by IvanSavenko — APPROVED** [review]

**Pull request merged**